packages/web-client/app/components/card-pay/deposit-workflow/transaction-amount/index.ts

```diff
@@ -129,64 +154,69 @@ class CardPayDepositWorkflowTransactionAmountComponent extends Component<Workflo
-      this.unlockTxnReceipt = transactionReceipt;
+      let transactionReceipt;
+      if (this.unlockTxnHash) {
+        transactionReceipt = yield this.layer1Network.resumeApprove(
+          this.unlockTxnHash
+        );
+      } else {
+        transactionReceipt = yield taskFor(
+          this.layer1Network.approveTask
+        ).perform(
+          this.amountAsBigNumber,
+          this.currentTokenSymbol,
+          (txnHash) => {
+            session.update('unlockTxnHash', txnHash);
+          }
+        );
+      }
+      session.update('unlockTxnReceipt', transactionReceipt);
     } catch (e) {
       console.error(e);
       this.errorMessage =
         'There was a problem unlocking your tokens for deposit. This may be due to a network issue, or perhaps you canceled the request in your wallet.';
-    } finally {
-      this.isUnlocking = false;
     }
   }
-  @action async deposit() {
-    this.errorMessage = '';
 
+  @task *depositTask(): TaskGenerator<void> {
+    this.errorMessage = '';
+    let session = this.args.workflowSession;
     try {
-      let layer2Address = this.layer2Network.walletInfo.firstAddress!;
-      this.isDepositing = true;
-      let layer2BlockHeightBeforeBridging =
-        await this.layer2Network.getBlockHeight();
-      this.args.workflowSession.update(
-        'layer2BlockHeightBeforeBridging',
-        layer2BlockHeightBeforeBridging
-      );
-      let transactionReceipt = await taskFor(
-        this.layer1Network.relayTokensTask
-      ).perform(
-        this.currentTokenSymbol,
-        layer2Address,
-        this.amountAsBigNumber,
-        (txnHash) => {
-          this.relayTokensTxnHash = txnHash;
-        }
-      );
-      this.relayTokensTxnReceipt = transactionReceipt;
-      this.args.workflowSession.update(
-        'relayTokensTxnReceipt',
-        transactionReceipt
-      );
-      this.args.workflowSession.update(
-        'depositedAmount',
-        this.amountAsBigNumber.toString()
-      );
+      let transactionReceipt;
+      if (this.relayTokensTxnHash) {
+        transactionReceipt = yield this.layer1Network.resumeRelayTokens(
+          this.relayTokensTxnHash
+        );
+      } else {
+        let layer2Address = this.layer2Network.walletInfo.firstAddress!;
+        let layer2BlockHeightBeforeBridging =
+          yield this.layer2Network.getBlockHeight();
+        session.update(
+          'layer2BlockHeightBeforeBridging',
+          layer2BlockHeightBeforeBridging
+        );
+        transactionReceipt = yield taskFor(
+          this.layer1Network.relayTokensTask
+        ).perform(
+          this.currentTokenSymbol,
+          layer2Address,
+          this.amountAsBigNumber,
+          (txnHash) => {
+            session.update('relayTokensTxnHash', txnHash);
+          }
+        );
+      }
+      session.update('relayTokensTxnReceipt', transactionReceipt);
+      session.update('depositedAmount', this.amountAsBigNumber.toString());
```

Comment:
Could use `updateMany` to update both at the same time. Actually, I think it's confusing we have both `update` and `updateMany`. Could just be one update method that accepts an object with keys and values (to remove `update` and rename `updateMany` to `update`)

Reply:
Left this along for now. I have an idea I'd like us to consider about this API.